### ISSUE TITLE
Partial download with network trouble marked complete rather than failed

### ISSUE THREAD

**Issue by shssoichiro** [downthemall/downthemall#495]
**Desktop (please complete the following information):**
 - OS: Windows 11
 - Browser and version: Firefox 122.0 64-bit
 - DownThemAll! version: 4.11.1

**Describe the bug**
Occasionally when downloading files with DownThemAll, I have noticed that the output file will be incomplete, typically due to a network issue on the remote end. However, DownThemAll will show these incomplete files as complete, despite knowing the expected filesize beforehand.

To explain a bit more clearly, here is the set of observations I have when this issue occurs:
- When the download begins and while downloading, the expected filesize is shown in DownThemAll manager.
- When a remote network issue causes the download to disconnect prior to finishing, DownThemAll marks the download as complete, and updates the file size in the manager to the size of the incomplete portion that was downloaded.

**To Reproduce**
1. Download a large file in DownThemAll
2. Have a network disconnection from the remote end mid-download
3. Observe that the download is marked as completed

To be fair, I'm not certain if it is all remote network issues or only some specific subset of issues that may be causing this to occur, so I apologize if it may be difficult to reproduce this.

**Expected behavior**
If the expected size is known, and the file received is smaller than that size, I would expect the download to be marked as failed rather than complete, so that I can be aware of the issue and resume or retry the download.

**Comment by nmaier** [comment]
Yeah, this is an unfortunate thing. It's the browser (Firefox in your case) that declares the download to be complete. One way to detect that is would be to compare the file size the server reported to the final file size the browser repots... the trouble with that is, however, that there are *tons* of webservers out there that report wrong file sizes (usually they would e.g. report the transfer encoding size, which is larger or smaller than the actual file). Additionally a lot of servers do not even report any file size at all (because e.g. the file is generated on the fly, e.g. all kinds of reports or pdfs or zips and therefore the size is not known beforehand).

What complicates this further is that if a download is interrupted in the way you describe, but the browser still marks it as complete, then there is no way to just override the browser and resume the download. The "best" one could do then is start the download from the beginning again - which then wouldn't much improve the situation at hand. These are the "new" constraints WebExtensions impose... Back in the "good old days" when extension authors had far more control, I remember spending many hours on getting the resumption code as "right" as possible, coming up with detection mechanisms and heuristics for all kind of common issues surrounding that. But that's not possible anymore, so I am a bit stumped at what exactly to do here, to be quite honest.
